mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,651 @@
+            metadata=cfg,  # Save all config as metadata in the checkpoint
+            model_name=cfg.logger.system_name,
+            **cfg.logger.checkpointing.save_args,  # Checkpoint args
+        )
+
+    max_episode_return = -jnp.inf
+    start_time = time.time()
+
+    # Main loop:
+    for eval_idx, t in enumerate(
+        range(steps_per_rollout, int(cfg.system.total_timesteps), steps_per_rollout)
+    ):
+        # Learn loop:
+        learner_state, (metrics, losses) = update(learner_state)
+        jax.block_until_ready(learner_state)
+
+        # Log:
+        # Add learn steps here because anakin steps per second is learn + act steps
+        # But we also want to make sure we're counting env steps correctly so
+        # learn steps is not included in the loop counter.
+        learn_steps = anakin_steps * cfg.system.epochs
+        sps = (t + learn_steps) / (time.time() - start_time)
+        eps = jax.numpy.maximum(
+            cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
+        )
+
+        final_metrics, ep_completed = episode_metrics.get_final_step_metrics(metrics)
+        loss_metrics = losses
+        logger.log({"step": t, "steps_per_second": sps, "epsilon": eps}, t, eval_idx, LogEvent.MISC)
+        if ep_completed:
+            logger.log(final_metrics, t, eval_idx, LogEvent.ACT)
+        logger.log(loss_metrics, t, eval_idx, LogEvent.TRAIN)
+
+        # Evaluate:
+        key, eval_key = jax.random.split(key)
+        eval_keys = jax.random.split(eval_key, cfg.arch.n_devices)
+        # essentially squeeze/unreplicate batch dim
+        eval_params = unreplicate_batch_dim(learner_state.params.online)
+        eval_output = evaluator(eval_params, eval_keys)
+        jax.block_until_ready(eval_output)
+
+        # Log:
+        episode_return = jnp.mean(eval_output.episode_metrics["episode_return"])
+        logger.log(eval_output.episode_metrics, t, eval_idx, LogEvent.EVAL)
+
+        # Save best actor params.
+        if cfg.arch.absolute_metric and max_episode_return <= episode_return:
+            best_params = copy.deepcopy(eval_params)
+            max_episode_return = episode_return
+
+        # Checkpoint:
+        if cfg.logger.checkpointing.save_model:
+            # Save checkpoint of learner state
+            unreplicated_learner_state = unreplicate_n_dims(learner_state)  # type: ignore
+            checkpointer.save(
+                timestep=t,
+                unreplicated_learner_state=unreplicated_learner_state,
+                episode_return=episode_return,
+            )
+
+    # Measure absolute metric.
+    if cfg.arch.absolute_metric:
+        eval_keys = jax.random.split(key, cfg.arch.n_devices)
+
+        eval_output = absolute_metric_evaluator(best_params, eval_keys)
+        jax.block_until_ready(eval_output)
+
+        logger.log(eval_output.episode_metrics, t, eval_idx, LogEvent.ABSOLUTE)
+
+    logger.stop()
+
+    return float(max_episode_return)
```

Comment:
This is a small bug. We only use this value for hydra sweeps and we don't want hydra sweeps to optimize for the best max return, we want them to optimize our final output value. Can you please see how PPO does this and update it. See [here](https://github.com/instadeepai/Mava/blob/develop/mava/systems/ppo/ff_ippo.py#L556)

Sorry I see I also made this mistake in SAC

Reply:
All good, thank you for the catch!!